libyul/ASTLabelRegistry.h

```diff
@@ -0,0 +1,106 @@
+#include <string>
+#include <string_view>
+#include <vector>
+
+namespace solidity::yul
+{
+
+/// Instances of the `ASTLabelRegistry` are immutable containers describing a labelling of nodes inside the AST.
+/// Each element of the AST that possesses a label has a `ASTLabelRegistry::LabelID`, with which the label can
+/// be queried in O(1).
+/// Preferred way of creating instances is via `ASTLabelRegistryBuilder` when parsing/importing and
+/// via `LabelIDDispenser` during/after optimization.
+/// Note: The id == 0 always corresponds to the empty label. Other ids can point at the label index 0, which means that
+/// they are unused. To check, whether an id is unused, the `unused` function can be used.
+/// Note: There is a special case of ghost ids, which are added during CFG construction.
+/// They do not directly correspond to elements of an AST but live inside the CFG. They are assigned the
+/// special `std::numeric_limits<LabelID>::max` label index. The labels themselves are - if required - generated from
+/// the CFG.
+class ASTLabelRegistry
+{
+public:
+	/// unsafe to use from a different registry instance, it is up to the user to safeguard against this
+	using LabelID = size_t;
+
+	ASTLabelRegistry();
+	ASTLabelRegistry(std::vector<std::string> _labels, std::vector<size_t> _idToLabelMapping);
+
+	std::string_view operator[](LabelID _id) const;
+
+	static bool constexpr empty(LabelID const _id) { return _id == emptyID(); }
+	static LabelID constexpr emptyID() { return 0; }
+	static size_t constexpr ghostLabelIndex() { return std::numeric_limits<size_t>::max(); }
+
+	bool unused(LabelID _id) const;
+	bool ghost(LabelID _id) const;
+	std::vector<std::string> const& labels() const { return m_labels; }
+	LabelID maxID() const;
+
+	size_t idToLabelIndex(LabelID _id) const;
+	/// this is a potentially expensive operation
+	std::optional<LabelID> findIdForLabel(std::string_view _label) const;
+
+private:
+	/// All Yul AST labels present in the registry.
+	/// Always contains at least an empty label which also serves as a null-state for non-contiguous ID ranges.
+	/// All items must be unique.
+	std::vector<std::string> m_labels;
+
+	/// Assignment of labels to `LabelID`s. Indices are `LabelID`s and values are indices into `m_labels`.
+	/// Every label except empty and ghost placeholder always has exactly one `LabelID` pointing at it.
+	std::vector<size_t> m_idToLabelMapping;
+};
+
+/// Produces instances of `ASTLabelRegistry`. Preferably used during parsing/importing.
+class ASTLabelRegistryBuilder
+{
+public:
+	ASTLabelRegistryBuilder();
+	/// Creates a new builder taking an already existing label registry into account.
+	/// Unused IDs are skipped, ghost IDs are recorded and inserted back into the newly built registry.
+	explicit ASTLabelRegistryBuilder(ASTLabelRegistry const& _existingRegistry);
+
+	ASTLabelRegistry::LabelID define(std::string_view _label);
+	ASTLabelRegistry::LabelID addGhost();
+	ASTLabelRegistry build() const;
```

Comment:
As mentioned in https://github.com/ethereum/solidity/pull/15823#discussion_r1954834363, please add a docstring explaining what gets copied. In particular that the unused labels do not carry over, but ghosts do.

Reply:
Ah, right. I got confused by the fact that we're filtering out unused IDs. But that's only because we don't explicitly store them.